cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll

```diff
@@ -1,313 +1,278 @@
 import cpp
 import AliasAnalysis
 import semmle.code.cpp.ir.internal.Overlap
+private import semmle.code.cpp.Print
 private import semmle.code.cpp.ir.implementation.unaliased_ssa.IR
 private import semmle.code.cpp.ir.internal.IntegerConstant as Ints
 private import semmle.code.cpp.ir.internal.IntegerInterval as Interval
 private import semmle.code.cpp.ir.internal.OperandTag
 
 private class IntValue = Ints::IntValue;
 
-private newtype TVirtualVariable =
-  TVirtualIRVariable(IRVariable var) {
-    not variableAddressEscapes(var)
-  } or
-  TUnknownVirtualVariable(FunctionIR f)
-
-private VirtualIRVariable getVirtualVariable(IRVariable var) {
-  result.getIRVariable() = var
-}
-
-private UnknownVirtualVariable getUnknownVirtualVariable(FunctionIR f) {
-  result.getEnclosingFunctionIR() = f
-}
-
-class VirtualVariable extends TVirtualVariable {
-  string toString() {
-    none()
-  }
-  
-  string getUniqueId() {
-    none()
-  }
-  
-  Type getType() {
-    none()
-  }
-}
-
-/**
- * A virtual variable representing a single non-escaped `IRVariable`.
- */
-class VirtualIRVariable extends VirtualVariable, TVirtualIRVariable {
-  IRVariable var;
-
-  VirtualIRVariable() {
-    this = TVirtualIRVariable(var)
-  }
-
-  override final string toString() {
-    result = var.toString()
-  }
-
-  final IRVariable getIRVariable() {
-    result = var
-  }
-
-  override final Type getType() {
-    result = var.getType()
-  }
-
-  override final string getUniqueId() {
-    result = var.getUniqueId()
-  }
-}
-
-/**
- * A virtual variable representing all escaped memory accessible by the function,
- * including escaped local variables.
- */
-class UnknownVirtualVariable extends VirtualVariable, TUnknownVirtualVariable {
-  FunctionIR f;
-  
-  UnknownVirtualVariable() {
-    this = TUnknownVirtualVariable(f)
-  }
-  
-  override final string toString() {
-    result = "UnknownVvar(" + f + ")"
-  }
-  
-  override final string getUniqueId() {
-    result = "UnknownVvar(" + f + ")"
-  }
-  
-  override final Type getType() {
-    result instanceof UnknownType
-  }
-  
-  final FunctionIR getEnclosingFunctionIR() {
-    result = f
-  }
-}
-
-private predicate hasResultMemoryAccess(Instruction instr, IRVariable var, IntValue startBitOffset,
+private predicate hasResultMemoryAccess(Instruction instr, IRVariable var, Type type, IntValue startBitOffset,
     IntValue endBitOffset) {
   resultPointsTo(instr.getResultAddressOperand().getDefinitionInstruction(), var, startBitOffset) and
+  type = instr.getResultType() and
   if exists(instr.getResultSize()) then
     endBitOffset = Ints::add(startBitOffset, Ints::mul(instr.getResultSize(), 8))
```

Comment:
Why `Ints::mul` and not just `*`? Also below.

Reply:
I think these should probably remain. The `startBitOffset` bound by `resultPointsTo()` could be the result of arbitrary pointer arithmetic. Overflow is unlikely, but possible.

Reply:
I agree that `Ints::add` should remain, but my question was about `Ints::mul`.

Reply:
Overflow in the `mul` calculation is even less likely, but still theoretically possible if the result of the instruction is a very large array type. Is there likely to be a significant performance cost from this use of `mul`? If so, it's probably OK to remove it, but if there's not likely to be much performance difference, I'd prefer to keep it for (pedantic) correctness.

Reply:
If `hasResultMemoryAccess` is fast, then it should be fine.